scene/resources/material.cpp

```diff
@@ -191,7 +193,10 @@ Variant ShaderMaterial::property_get_revert(const String &p_name) {
 
 void ShaderMaterial::set_shader(const Ref<Shader> &p_shader) {
 
-	if (shader.is_valid()) {
+	// Only connect/disconnect the signal when running in the editor.
+	// This can be a slow operation, and `_change_notify()` (which is called by `_shader_changed()`)
+	// does nothing in non-editor builds anyway. See GH-34741 for details.
```

Comment:
Maybe add a comment in `_change_notify()` too to mention that it's only connected in the editor - in case we want to add more logic there in the future that would be supposed to run in-game too.